File: backend/src/tools/sandbox/multi_edit/multi_edit.py
"""Multi-edit tool: apply an ordered batch of edits to one file atomically."""

from __future__ import annotations

from sandbox._shared.models import Intent

from pydantic import BaseModel, ConfigDict, Field

import sandbox.api as sandbox_api
from sandbox.api import EditFileRequest, SearchReplaceEdit
from tools._framework.core.base import ToolExecutionContextService, ToolResult
from tools._framework.core.decorator import tool
from tools.sandbox._lib.tool_context import (
    sandbox_audit_kwargs_from_tool_context,
    sandbox_caller_from_tool_context,
    resolve_tool_sandbox_path,
    sandbox_id_or_missing_error_result,
)
from tools.sandbox._lib.mutation_result import project_file_mutation
from .prompt import get_multi_edit_description


class MultiEditOp(BaseModel):
    model_config = ConfigDict(extra="forbid")

    old_text: str = Field(..., description="Exact text to find.")
    new_text: str = Field(default="", description="Replacement text.")
    replace_all: bool = Field(
        default=False,
        description="Replace every occurrence of `old_text` instead of requiring a unique match.",
    )


class MultiEditInput(BaseModel):
    model_config = ConfigDict(extra="forbid")

    file_path: str = Field(..., description="Repo-relative or sandbox-root file path.")
    edits: list[MultiEditOp] = Field(
        ...,
        description=(
            "Ordered edits applied sequentially against evolving content "
            "(edit N sees edit N-1's result); all-or-nothing."
        ),
    )
    description: str = Field(
        default="",
        description="Optional short note about the edits.",
    )


class MultiEditOutput(BaseModel):
    cwd: str = Field(..., description="Current sandbox working directory.")
    file_path: str = Field(..., description="Resolved file path that was edited.")
    status: str = Field(..., description="Edit result: edited, aborted_version, or failed.")
    changed_paths: list[str] = Field(default_factory=list, description="Files changed by the edits.")
    changed_path_kinds: dict[str, str] = Field(
        default_factory=dict,
        description="Changed paths keyed to write/delete/symlink/opaque_dir.",
    )
    mutation_source: str = Field(default="", description="Mutation source tag.")
    conflict_reason: str | None = Field(
        default=None, description="Conflict reason when the edits failed."
    )
    error: dict[str, object] = Field(default_factory=dict, description="Typed error payload.")
    applied_edits: int = Field(
        default=0,
        description="Number of edits applied (not occurrence count).",
    )


@tool(
    name="multi_edit",
    description=get_multi_edit_description(),
    short_description="Apply an ordered batch of edits to one file atomically.",
    input_model=MultiEditInput,
    output_model=MultiEditOutput,
    intent=Intent.WRITE_ALLOWED,
)
async def multi_edit(
    file_path: str,
    edits: list[dict[str, object]],
    description: str = "",
    *,
    context: ToolExecutionContextService,
) -> ToolResult:
    """Apply an ordered batch of edits to one file."""
    file_path = resolve_tool_sandbox_path(file_path, context)

    if not edits:
        return ToolResult(
            output="Provide at least one edit in `edits`.",
            is_error=True,
        )

    sandbox_id, sandbox_id_error = sandbox_id_or_missing_error_result(context)
    if sandbox_id_error is not None:
        return sandbox_id_error

    search_replace_edits = tuple(
        SearchReplaceEdit(
            old_text=str(op.get("old_text", "")),
            new_text=str(op.get("new_text", "")),
            replace_all=bool(op.get("replace_all", False)),
        )
        for op in edits
    )

    result = await sandbox_api.edit_file(
        sandbox_id,
        EditFileRequest(
            path=file_path,
            edits=search_replace_edits,
            caller=sandbox_caller_from_tool_context(context),
            description=description or f"multi-edit {file_path}",
        ),
        **sandbox_audit_kwargs_from_tool_context(context),
    )

    return project_file_mutation(
        result,
        success_status="edited",
        file_path=file_path,
        success_extra={"applied_edits": len(search_replace_edits)},
        context=context,
    )


__all__ = ["multi_edit"]

File: agent-core/crates/eos-tools/src/tools/sandbox/multi_edit.rs
use async_trait::async_trait;
use eos_sandbox_api::{EditFileRequest, SearchReplaceEdit};
use eos_types::JsonObject;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::core::error::ToolError;
use crate::core::metadata::ExecutionMetadata;
use crate::core::name::ToolName;
use crate::core::result::ToolResult;
use crate::runtime::execution::parse_input;
use crate::runtime::executor::ToolExecutor;

use super::super::SandboxToolService;
use super::lib::{default_empty, default_false, edit_output, request_base, resolve_path};

#[derive(Debug, Deserialize, Serialize, JsonSchema)]
pub(super) struct MultiEditOp {
    old_text: String,
    #[serde(default = "default_empty")]
    new_text: String,
    #[serde(default = "default_false")]
    replace_all: bool,
}

#[derive(Debug, Deserialize, Serialize, JsonSchema)]
pub(super) struct MultiEditInput {
    file_path: String,
    edits: Vec<MultiEditOp>,
    #[serde(default = "default_empty")]
    description: String,
}

pub(super) struct MultiEdit {
    service: SandboxToolService,
}

impl MultiEdit {
    pub(super) fn new(service: SandboxToolService) -> Self {
        Self { service }
    }
}

#[async_trait]
impl ToolExecutor for MultiEdit {
    async fn execute(
        &self,
        input: &JsonObject,
        ctx: &ExecutionMetadata,
    ) -> Result<ToolResult, ToolError> {
        let parsed: MultiEditInput = match parse_input(ToolName::MultiEdit, input) {
            Ok(v) => v,
            Err(err) => return Ok(err),
        };
        if parsed.edits.is_empty() {
            return Ok(ToolResult::error("Provide at least one edit in `edits`."));
        }
        let path = resolve_path(ctx, &parsed.file_path);
        let sandbox_id = ctx.require_sandbox_id()?;
        let count = parsed.edits.len() as u64;
        let description = if parsed.description.is_empty() {
            format!("multi-edit {path}")
        } else {
            parsed.description.clone()
        };
        let request = EditFileRequest {
            base: request_base(ctx, &description)?,
            path: path.clone(),
            edits: parsed
                .edits
                .into_iter()
                .map(|op| SearchReplaceEdit {
                    old_text: op.old_text,
                    new_text: op.new_text,
                    replace_all: op.replace_all,
                })
                .collect(),
        };
        let result = match eos_sandbox_api::edit_file(
            &*self.service.transport,
            sandbox_id,
            &request,
        )
        .await
        {
            Ok(result) => result,
            Err(err) => return Ok(ToolResult::error(err.to_string())),
        };
        // multi_edit reports the count of edits submitted (not result.applied_edits).
        let applied = if result.base.success { count } else { 0 };
        Ok(edit_output(
            ctx,
            path,
            &result.base,
            result.changed_path_kinds,
            result.mutation_source,
            applied,
        ))
    }
}
